src/lm_pub_quiz/integration.py

```diff
@@ -0,0 +1,43 @@
+import numpy as np
+from transformers import TrainerCallback
+
+from lm_pub_quiz.data import Dataset
+from lm_pub_quiz.evaluator import Evaluator
+
+
+class PubQuizCallback(TrainerCallback):
+    """This callback handels on the fly BEAR evaluation for the huggingface Trainer class"""
+    def __init__(self, trainer, dataset_path, metrics=["accuracy"], model_type=None, save_path=None, batch_size=32, template=0):
+        self.trainer = trainer
+        self.model_type = model_type
+        self.save_path = save_path
+        self.batch_size = batch_size
+        self.template = template
+        if "accuracy" in metrics and "num_instances" not in metrics:
+            metrics.append("num_instances")
+        self.metrics = metrics
+        # include default loading of dataset from url?
+        self.dataset = Dataset.from_path(dataset_path)
+
+        self.dataset = self.dataset.filter_subset({'P6': list(range(10)), 'P19': list(range(10))})
+
+    def on_evaluate(self, args, state, control, **kwargs):
+        evaluator = Evaluator.from_model(
+            kwargs['model'],
+            model_type=self.model_type,
+            tokenizer=kwargs['tokenizer'],
+            device=kwargs['model'].device,
+        )
+        result = evaluator.evaluate_dataset(self.dataset, template_index=self.template, batch_size=self.batch_size)
+        if self.save_path:
+            result.save(f"{self.save_path}/{state.epoch}_bear_results")
+
+        bear_metrics = result.get_metrics(self.metrics)
+        weighted_accuracy = np.average(bear_metrics.accuracy, weights=bear_metrics.num_instances)
+
+        # TODO: Include additional metrics
+        metrics_data = {
+            'eval_bear_score': weighted_accuracy
```

Comment:
Maybe we can retrieve the relevant metrics from `self.metrics` and report all of the ones that have been passed.

Reply:
Ideally, we would also have an option to retrieve all combinations from the resulting dataframe, if the user wants to accumulate over the domains for example. Potentially, this would require multiple calls to `get_metrics` if they would like to track different dimensions (e.g. cardinality + domains).